src/Surfnet/Stepup/Identity/Identity.php

```diff
@@ -235,24 +238,74 @@ public function verifyEmail($verificationNonce)
         $secondFactorToVerify->verifyEmail();
     }
 
-    public function vetSecondFactor($registrationCode, $secondFactorIdentifier, $documentNumber, $identityVerified)
-    {
-        foreach ($this->verifiedSecondFactors as $secondFactor) {
-            if (!$secondFactor->wouldBeVettedBy(
-                $registrationCode,
-                $secondFactorIdentifier,
-                $documentNumber,
-                $identityVerified
-            )) {
-                continue;
-            }
+    public function vetSecondFactor(
+        IdentityApi $identity,
+        SecondFactorId $secondFactorId,
+        $secondFactorIdentifier,
+        $registrationCode,
+        $documentNumber,
+        $identityVerified
+    ) {
+        /** @var VettedSecondFactor|false $secondFactorWithHighestLoa */
+        $secondFactorWithHighestLoa = array_reduce(
+            $this->vettedSecondFactors->toArray(),
+            function (VettedSecondFactor $carry, VettedSecondFactor $item) {
+                return $carry->hasEqualOrHigherLoaComparedTo($item) ? $carry : $item;
+            },
+            $this->vettedSecondFactors->first()
+        );
 
-            $secondFactor->vet($registrationCode, $secondFactorIdentifier, $documentNumber, $identityVerified);
-            return;
+        $verifiedSecondFactorHasEqualOrLowerLoaComparedTo =
+            $identity->verifiedSecondFactorHasEqualOrLowerLoaComparedTo(
+                $secondFactorId,
+                $secondFactorWithHighestLoa
+            );
+
+        if (!$verifiedSecondFactorHasEqualOrLowerLoaComparedTo) {
+            throw new DomainException('Authority does not have the required LoA to vet the identity\'s second factor');
+        }
```

Comment:
Wall of Text inc!

TL;DR: not sure how I feel about this all, but it's just a feeling. Have a different implementation below, how would you feel about something like that?

---

In all honesty, I'm not sure how I feel about these lines.... I've been hesitant about whether or not I should say something since it's just a feeling and it may well be that it's since I'm reading it on github, but these lines (and their consequences) are very hard to read/grok; amongst things due to the long variable and method names and the triple dispatch behind it. 
It kinda feels like that it could be simpler by just requesting the Verified Second Factor from the Identity and checking the LoA's of both... something like:

```php
$secondFactorToBeVetted = $identityToBeVetted->getVerifiedSecondFactor($sfid);
if (!$mySecondFactor->hasLoaHigherOrEqualTo($secondFactorToBeVetted) {
    throw new DomainException();
}
```
Which is somewhat analogous to what is happening in reality... 
The upside of something like that would also be that you get to call `$secondFactorToBeVetted->vet(/* params */)` albeit that might well be a bridge too far...

Also, `$identity->complyWithVettingOfSecondFactor` could also be `$identity->complyWithVetting()` as the Identity itself is vetted as well :wink:

---

With the renaming and SecondFactorCollection this would become something like:
```php
// when we have to check based on current LoA (future) this might become 
// $this->vettedSecondFactors->get('xxx');
$mySecondFactor = $this->vettedSecondFactors->getSecondFactorWithHighestLoa();
$registrantSecondFactor = $registrant->getVerifiedSecondFactor($registrantSecondFactorId);

if (!$mySecondFactor->hasLoaHigherThanOrEqualTo($registrantSecondFactor()) {
    throw new DomainException('naughty boy!');
}

$registrant->complyWithVetting(/* params */);
```

Thoughts?